### PR TITLE
Add Starcounter.Validation

### PR DESCRIPTION
**Pull request by Mihaiii**

### CODE REVIEW

**Comment by mmnosek** [comment]
@joozek78 this waits for your review.

**Comment by Mihaiii** [comment]
I have to update the changes in accordance with the latest SC.Validation (will have to be sure first that every other app is using the same version of SC.Validation).

**Comment by Mihaiii** [comment]
Related to https://github.com/Starcounter/HeadsOmni/issues/332

**Comment by Mihaiii** [comment]
I will ping @joozek78 when I'm finised with the updates, please don't review before that.

**Comment by Mihaiii** [comment]
Ok, current situation is like this:
`ValidatorBuilder` no longer has a constructor that takes 0 args. Therefore, `AddStarcounterValidation()` and DI are enforced, but the problem is that current latest SC mapping version doesn't support constructors that take args, therefore `IInitPageWithDependencies` has to be used and here I'm running into some code behind json auto generated code error.

useradmin

This is blocked until we have a SC mapping version that can have SC.Json constructors that accept args (from what I understood this is already fixed on level1's `develop`).

**Comment by Mihaiii** [comment]
@joozek78 Please have another look. :)

File: src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            var value = (string)val;
+
+            if (value == null || string.IsNullOrEmpty(value.Trim()))
+            {
+                return ValidationResult.Success;
+            }
+
+            var systemUserRepository = validationContext.GetRequiredService<ISystemUserRepository>();
+
+            if (systemUserRepository.GetSystemUserByUserName(value) == null)
```
Comment (joozek78) [inline src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs]: why not split this into two attributes? `UniqueUsernameAttribute` and `NotAMemberAttribute`

File: src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            {
+                return new ValidationResult("User not found.");
+            }
+
+            var property = validationContext.ObjectType.GetProperty(_other);
```
Comment (joozek78) [inline src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs]: why accept the property as an argument? You could just cast `validationContext.ObjectInstance` to the view-model type and get the property directly. If you want to reuse over multiple view-models you could introduce `IUserViewModel` which will have this property

File: src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs (hunk 1 above)
Comment (joozek78) [reply src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs]: then also you can use `FormatErrorMessage`

File: src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+            foreach (var member in group.Members)
+            {
+                if (member.Username.Equals(value, StringComparison.InvariantCultureIgnoreCase))
+                {
+                    return new ValidationResult("User is already member of the group.");
+                }
+            }
+
+            return null;
```
Comment (joozek78) [inline src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs]: use `ValidationResult.Success` instead. It's more readable

File: src/UserAdmin/Helpers/ValidateNameAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                );
+            }
+            var otherValue = property.GetValue(validationContext.ObjectInstance, null) as string;
+            var systemUserGroupRepository = validationContext.GetRequiredService<ISystemUserGroupRepository>();
+            var data = systemUserGroupRepository.GetSystemUserGroupByKey(otherValue);
+            var result = systemUserGroupRepository.GetFirstSystemUserGroupWithSameName(value, data);
+
+            if (result != null)
+            {
+                return new ValidationResult("System group with that name already exists.");
```
Comment (joozek78) [inline src/UserAdmin/Helpers/ValidateNameAttribute.cs]: please use `FormatErrorMessage`, as described [here](https://github.com/Starcounter/Starcounter.Validation#custom-validation-rules)

File: src/UserAdmin/Helpers/ValidateSelectedBasedOnGroupIDAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Starcounter;
+using UserAdmin.Repositories;
+
+namespace UserAdmin.Helpers
+{
+    public class ValidateSelectedBasedOnGroupIDAttribute : ValidationAttribute
+    {
+        private readonly string _other;
```
Comment (joozek78) [inline src/UserAdmin/Helpers/ValidateSelectedBasedOnGroupIDAttribute.cs]: see my previous comments about accepting the property argument and formatting error message

File: src/UserAdmin/ViewModels/Administrator/EditPersonPage.json.cs [hunk 6]
```diff
@@ -1,63 +1,55 @@
...
-        }
-
-        #region Validate Properties (Create Property Feedbacks)
-
-        protected void AssurePropertyFeedback_Name(string propertyName, string value)
-        {
-            string message;
-            this.Validate_Name(value, out message);
-            if (message != null)
+            if (!_validator.Validate(nameof(FirstName), action.Value))
```
Comment (joozek78) [inline src/UserAdmin/ViewModels/Administrator/EditPersonPage.json.cs]: why have `if` if both branches do the same (return)?

Comment (joozek78) [reply src/UserAdmin/ViewModels/Administrator/EditPersonPage.json.cs]: am I missing something?

File: src/UserAdmin/ViewModels/Administrator/EditUserGroupPage.json.cs [hunk 7]
```diff
@@ -60,39 +86,19 @@ public IEnumerable Groups
...
-                        this.Message = "Circular referenced system groups are not allowd";
-                        action.Cancel();
-                        action.Cancelled = true;
-                        this.SelectedBasedOnGroupID = action.OldValue;
-                        return;
-                    }
-
-                    group = group.Parent;
-                }
+                action.Cancel();
```
Comment (joozek78) [inline src/UserAdmin/ViewModels/Administrator/EditUserGroupPage.json.cs]: I think those do the same thing

File: src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs (hunk 2 above)
Comment (Mihaiii) [reply src/UserAdmin/Helpers/ValidateAddUserNameAttribute.cs]: Yes, I noticed that that's the way MS recommend it - why is that? My problem with it is that in this case the attribute class will only work for one vm class, when I see the attribute as being general.

Anyway, I'll address soon.

**Comment by Mihaiii** [comment]
@joozek78 Addressed. Please have another look.

File: src/UserAdmin/Helpers/ExistingUserAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+            if (value == null || string.IsNullOrEmpty(value.Trim()))
+            {
+                return ValidationResult.Success;
+            }
+
+            var systemUserRepository = validationContext.GetRequiredService<ISystemUserRepository>();
+
+            if (systemUserRepository.GetSystemUserByUserName(value) == null)
+            {
+                return new ValidationResult(FormatErrorMessage("User not found."));
```
Comment (joozek78) [inline src/UserAdmin/Helpers/ExistingUserAttribute.cs]: have you tested that? I believe the argument to `FormatErrorMessage` should be the name of the invalid property. So, in most cases, `validationContext.DisplayName`

File: src/UserAdmin/Helpers/NotAMemberAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+            var key = ((EditUserGroupPage)validationContext.ObjectInstance).Data.GetObjectID();
+
+            // Check if user is alread a member of this sytem user group
+            var systemUserGroupRepository = validationContext.GetRequiredService<ISystemUserGroupRepository>();
+            var group = systemUserGroupRepository.GetSystemUserGroupByKey(key);
+
+            foreach (var member in group.Members)
+            {
+                if (member.Username.Equals(value, StringComparison.InvariantCultureIgnoreCase))
```
Comment (joozek78) [inline src/UserAdmin/Helpers/NotAMemberAttribute.cs]: is comparing by username correct? shouldn't you compare by objectId?

**Comment by Mihaiii** [comment]
@joozek78 This is ready for a re-review.
Thank you for the help! :)

**Pull request merged**